compiler/transforms/reuse_memref_space.py

```diff
@@ -1,72 +1,104 @@
-from collections.abc import Sequence
-
 from xdsl.context import MLContext
 from xdsl.dialects import affine, arith, builtin, memref, scf
 from xdsl.dialects.builtin import IndexType
-from xdsl.ir import Block, Operation, OpResult, SSAValue
+from xdsl.ir import Block, Operation, SSAValue
 from xdsl.ir.affine import AffineConstantExpr
 from xdsl.passes import ModulePass
 from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
     PatternRewriter,
     PatternRewriteWalker,
     RewritePattern,
     op_type_rewrite_pattern,
 )
 from xdsl.rewriter import InsertPoint
+from xdsl.traits import Pure
+
+
+def find_parent_for_loop(op: Operation) -> scf.For | None:
+    """
+    Find the parent for-loop of the operation.
+    If no for-loop is found, return None.
+    """
+    op = op.parent_op()
+    if op is None:
+        return None
+    while not isinstance(op, scf.For):
+        if op.parent_op() is None:
+            return None
+        op = op.parent_op()
+    return op
+
+
+def is_in_loop(op: Operation) -> bool:
+    """
+    Check if the operation is inside a loop.
+    """
+    return find_parent_for_loop(op) is not None
+
+
+def defined_outside_loop(op: Operation) -> bool:
+    """
+    Check if all operands of the operation are defined outside the loop.
+    """
+    for operand in op.operands:
+        if find_parent_for_loop(operand.owner) is find_parent_for_loop(op):
+            return False
+    return True
 
 
-class MoveMemrefAllocations(RewritePattern):
+class LowerPureOperations(RewritePattern):
```

Comment:
I think you are referring to loop hoisting, or loop-invariant-code-motion. Lowering is something different.
```suggestion
class LoopHoistPureOperations(RewritePattern):
```